### DISCUSSION TITLE
Support for python code execution inside a documentation

### DISCUSSION THREAD

**Discussion by petrhrobar** [mkdocs/mkdocs#2835] [General]
Is there currently a feature that allows to source some python function and output the function's output?

Suppose I have a function:

```python

def func(name: str, age: int):
      """
       This function prints your name and age.

        ```python
        >>>func("Thomas", 15)
        >>>hey Thomas who is 15.
        ```
        
         """
        return f"hey {name} who is {age}."
 ```      


Now docstring provides an example and it would be rendered. However, the function example is kind of "hardcoded" when I change the function and not the docstring, the example output does not change as it is static.

Would it be possible to also some e.g.:

.. execute_code::

from src.example import func
func("Petr", 101)


This would always change the output as the function is being executed. 
I have been looking at sphinx feature [HERE](https://github.com/jpsenior/sphinx-execute-code)

**Comment by pawamoy** [comment]
Take a look at [Markdown Exec](https://github.com/pawamoy/markdown-exec) (I'm the author). It should be possible to use the extension in docstrings code blocks as well, given you are using [mkdocstrings](https://github.com/mkdocstrings/mkdocstrings) to inject them in your pages, though I did not test that.

Example:

```python
def func(name: str, age: int):
    """
    This function prints your name and age.

    ```python exec="yes" source="tabbed-left"
    from module import func  # you still have to import it :/
    print(func("Thomas", 15))
    ```
    """
    return f"hey {name} who is {age}."
```

The `source="tabbed-left"` option requires that you also use the `pymdownx.tabbed` extension.

UPDATE: updated the example.

**Reply by facelessuser** [reply]
@pawamoy, An Interesting plugin, looks useful 🙂 . I've also been considering packaging up a custom fence for SuperFences that does general execution of code within a code block as well, though my approach is more geared towards rendering code as if it was run in a Python repl.

My current prototype is a bit more advanced than I'd ship as it can run during normal Markdown processing, but it can also be packaged up in the browser and run live with Pyodide to let people tinker with the example: https://facelessuser.github.io/coloraide/.

So, right now it is very specific for that project's needs, but I imagine if I pulled out the color stuff, and didn't worry about auto-packaging it in the browser, it'd be a pretty good solution for static repl type display of code.

**Reply by pawamoy** [reply]
Whew, that's nice! I'd definitely be interested in a pymdown extension that augments pycon code blocks into browser-python repls :slightly_smiling_face:

**Reply by facelessuser** [reply]
Yeah, to load it up automatically in a browser would more likely need to be a MkDocs plugin as you'd have to inject the JS and additional CSS into the path for toggling between source and output and the logic to actually execute the code in the browser.

What I have now takes normal python code, executes it, and displays it as if it was run in a repl. We execute the code and get the AST output  and construct the repl format.

I was thinking of starting out with a pymdown custom SuperFence extension (as that is how I implent it in ColorAide currently).

Then, if desired, a MkDocs plugin could be built on top of that functionality. Though you'd have to provide some method of telling the plugin which packages to include in the browser. I imagine a list specifying packages already supported in Pyodide and pure Python wheels would be sufficient. I imagine it could get quite heavy if your package requires a ton of dependencies though 🙃. Oh, and if you want to include the latest release of your current project, you probably have to build the wheel and package it as well.

The live browser part gets a little complicated because of all the moving parts, but basic code -> repl is a good place to start as I already have that working and it is should be easy to package.

**Reply by pawamoy** [reply]
I wanted to take a look at Brython again, to see how it compares to Pyodide, and found this haha: https://brython.info/gallery/markdown_exec.html

**Reply by facelessuser** [reply]
Yeah, I load and execute Markdown as well. If you use the markdown button (top right) of this page, you can load Markdown from raw gist links or raw Github repo pages: https://facelessuser.github.io/coloraide/playground/. It's so I could play with examples, sort of like a limited notebook.

**Reply by facelessuser** [reply]
An example of it dynamically rendering the README.md file from a raw link: https://facelessuser.github.io/coloraide/playground/?notebook=https%3A%2F%2Fraw.githubusercontent.com%2Ffacelessuser%2Fcoloraide%2Fmaster%2FREADME.md

**Reply by facelessuser** [reply]
Finally got a custom fence up. https://gist.github.com/facelessuser/53fa4d93f27c252fda813b5e0ba7325c

@pawamoy I know you maintain https://github.com/pawamoy/markdown-exec. I saw you have some form of Pycon support, but I think it is very different. You may find this comment useful as it has examples: https://github.com/facelessuser/pymdown-extensions/issues/1690#issuecomment-1445620049.

I hesitate to add something directly like this to Pymdown Extensions, but it may make sense in something like `markdown-exec`? Either way, figured I'd post this as some may find it useful. I'll probably be trialing it in some other projects.

**Reply by pawamoy** [reply]
In Markdown Exec, the output of pycon code blocks is concatenated at the end. Your solution is better in that the output is inserted where it should be, and not concatenated at the end :+1: So I could definitely reuse some of your code to improve the UX. But I'm not sure how I'd use the custom fence itself. As you mentioned it, our goals are very different, so the custom fence might not be a good fit in Markdown Exec.

**Reply by facelessuser** [reply]
That's fair. My approach is more aimed at dropping in normal python code, but then executing it and displaying it as an example in pycon. As it was relevant to the discussion, and I had mentioned I would probably do this, I figured I make a callback here. Regardless, I have use for it in some other projects, I'll figure out what to do with it moving forward.

**Reply by pawamoy** [reply]
Yes of course, thanks for updating this discussion and mentioning me :slightly_smiling_face:
